File: apps/mobile/utils/time.ts
export function formatMessageTime(dateInput?: string | number): string {
  if (!dateInput) return "";
  const date = new Date(dateInput);
  if (Number.isNaN(date.getTime())) return "";
  const hours = date.getHours();
  const minutes = date.getMinutes();
  const ampm = hours >= 12 ? "PM" : "AM";
  const h12 = hours % 12 === 0 ? 12 : hours % 12;
  return `${h12}:${minutes.toString().padStart(2, "0")} ${ampm}`;
}

export function formatRelativeTime(dateInput?: string | number): string {
  if (!dateInput) return "";
  const date = new Date(dateInput);
  const now = new Date();
  const diffSec = Math.floor((now.getTime() - date.getTime()) / 1000);
  if (isNaN(diffSec) || diffSec < 0) return "";
  if (diffSec < 60) return "just now";
  const diffMin = Math.floor(diffSec / 60);
  if (diffMin < 60) return `${diffMin}m ago`;
  const diffHr = Math.floor(diffMin / 60);
  if (diffHr < 24) return `${diffHr}h ago`;
  const diffDays = Math.floor(diffHr / 24);
  return `${diffDays}d ago`;
}

File: apps/android/app/src/main/kotlin/com/console/mobile/core/util/Time.kt
package com.console.mobile.core.util

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/** Port of apps/mobile/utils/time.ts */
fun formatMessageTime(dateInput: Any?): String {
    if (dateInput == null) return ""
    val ms = when (dateInput) {
        is Number -> dateInput.toLong()
        is String -> dateInput.toLongOrNull() ?: try { Date.parse(dateInput) } catch (_: Exception) { return "" }
        else -> return ""
    }
    return try {
        SimpleDateFormat("h:mm a", Locale.US).format(Date(ms))
    } catch (_: Exception) {
        ""
    }
}

fun formatRelativeTime(dateInput: Any?, nowMs: Long = System.currentTimeMillis()): String {
    if (dateInput == null) return ""
    val ms = when (dateInput) {
        is Number -> dateInput.toLong()
        is String -> dateInput.toLongOrNull() ?: return ""
        else -> return ""
    }
    val diffSec = (nowMs - ms) / 1000
    if (diffSec < 0) return ""
    if (diffSec < 60) return "just now"
    val diffMin = diffSec / 60
    if (diffMin < 60) return "${diffMin}m ago"
    val diffHr = diffMin / 60
    if (diffHr < 24) return "${diffHr}h ago"
    return "${diffHr / 24}d ago"
}
